File: Androidparty/app/src/main/java/com/tzemaitis/androidparty/MainActivity.java
package com.tzemaitis.androidparty;

import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity implements LoginFragment.LoginFragmentListener,
                                                               LoadingFragment.LoadingFragmentListener,
                                                               ServerListFragment.ServerListFragmentListener {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        transaction.add(R.id.container, LoginFragment.newInstance());
        transaction.commit();
    }


    @Override
    public void onLogin(AuthenticationToken token) {
        changeFragment(LoadingFragment.newInstance(token.getToken()));
    }


    @Override
    public void onLoaded(ArrayList<Server> servers) {
        changeFragment(ServerListFragment.newInstance(servers));
    }

    @Override
    public void onLoadFailed(String message) {

        // TODO failed to get server list and returned back to login fragment.
        // We have to updateLoginFragment with error message to inform user whats happened

        changeFragment(LoginFragment.newInstance());
    }

    @Override
    public void onLogOut() {
        changeFragment(LoginFragment.newInstance());
    }

    private void changeFragment(Fragment fragment) {
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        transaction.replace(R.id.container, fragment);

        // TODO add custom animation for fragment transition
        // transaction.setCustomAnimations()

        transaction.commit();
    }
}

File: app/src/main/java/com/example/androidparty/view/MainActivity.kt
package com.example.androidparty.view

import android.os.Bundle
import androidx.fragment.app.FragmentManager
import com.example.androidparty.R
import dagger.android.support.DaggerAppCompatActivity

class MainActivity : DaggerAppCompatActivity(), FragmentsNavigationListener {

    lateinit var fragmentManager: FragmentManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        fragmentManager = this.supportFragmentManager
        val loginFragment = LoginFragment.newInstance(this)
        val fragmentTransaction = fragmentManager.beginTransaction()
        fragmentTransaction.replace(R.id.fragment_place, loginFragment)
        fragmentTransaction.commit()
    }

    override fun onLoginClicked() {
        val serverListFragment = ServerListFragment.newInstance(this)
        val fragmentTransaction = fragmentManager.beginTransaction()
        fragmentTransaction.replace(R.id.fragment_place, serverListFragment)
        fragmentTransaction.commit()
    }

    override fun onLogoutClicked() {
        val loginFragment = LoginFragment.newInstance(this)
        val fragmentTransaction = fragmentManager.beginTransaction()
        fragmentTransaction.replace(R.id.fragment_place, loginFragment)
        fragmentTransaction.commit()
    }

}
